llama.cpp

```diff
@@ -3373,22 +3371,25 @@ struct llm_tokenizer_bpe {
     llm_bigram_bpe::queue work_queue;
 };
 
-static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab & vocab, const std::string & raw_text, bool bos) {
+static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab & vocab, std::string raw_text, bool bos) {
     std::vector<llama_vocab::id> output;
 
+    if (bos && vocab.special_bos_id != -1) {
+        output.push_back(vocab.special_bos_id);
+    }
+
     if (raw_text.empty()) {
         return output;
     }
 
-    if (bos && vocab.special_bos_id != -1) {
-        output.push_back(vocab.special_bos_id);
-    }
+    raw_text = " " + raw_text;
```

Comment:
We should have this optional like the `add_bos` parameter.

Suggestion:
`static std::vector<llama_vocab::id> llama_tokenize_internal(const llama_vocab & vocab, std::string raw_text, bool prepend_bos, bool prepend_whitespace) {`

Reply:
I moved it to the SPM branch. It is not optional - without this, tokenization is incorrect.
This explains why it was added in the first place, but I now realize we always have to prefix whatever input we pass to the tokenizer, so there is no point to have this in the user code and risk someone forgetting to do it.

I have verified this using the new Python `test-tokenizer-0.py` by comparing the results with `test-tokenizer-0.cpp`